### ISSUE TITLE
False positive error E0308 due to type confusion from a glob use and use statement ordering

### ISSUE THREAD

**Issue by Architector4** [rust-lang/rust-analyzer#21679]
**rust-analyzer version**: `rust-analyzer 1 (7cb789d982 2026-02-01)` (from `rust-analyzer --version`)

**rustc version**: `rustc 1.93.1 (01f6ddf75 2026-02-11) (Arch Linux rust 1:1.93.1-1.1)`

**editor or extension**: happens with any including commandline

**relevant settings**: defaults

**repository link (if public, optional)**: https://archlinux.org/packages/extra/x86_64/rust-analyzer/

**code snippet to reproduce**: `cargo new something`, then put this into `src/main.rs`
```rust
pub use std::cmp::*;

pub mod evil {
    pub struct Ordering(pub i32);
}

pub mod oblivious {
    use crate::Ordering;

    pub fn what() -> Ordering {
        Ordering(2)
    }
}
pub use evil::Ordering;


fn main() {
    println!("Hello, world!");
}

```

This snippet causes a false positive error from rust-analyzer that cargo check doesn't report:

```
architector4@AGOGUS:/tmp/something$ cargo check
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
architector4@AGOGUS:/tmp/something$ rust-analyzer diagnostics .
at crate ass, file /tmp/something/src/main.rs: Error RustcHardError("E0308") from LineCol { line: 10, col: 8 } to LineCol { line: 10, col: 19 }: expected Ordering, found Ordering

diagnostic scan complete

Error: diagnostic error detected

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: start_thread
             at /usr/src/debug/glibc/glibc/nptl/pthread_create.c:454:8
   5: __GI___clone3
             at /usr/src/debug/glibc/glibc/misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:78:0
```

Not sure if that diagnostic error error or stack backtrace should be there normally, but I guess that's there too

Note the position of that last `pub use` being _after_ the two modules were defined. Moving it before `oblivious` resolves the false positive.

(I sprinkled `pub` everywhere to get rid of "unused" warnings, but that doesn't seem to be necessary to reproduce the bug.)

**Comment by ChayimFriedman2** [comment]
Seems exactly like https://github.com/rust-lang/rust-analyzer/issues/19224, which was allegedly fixed (https://github.com/rust-lang/rust-analyzer/pull/21561) - CC @Veykril.

**Comment by Architector4** [comment]
Ooh, yeah, that does look very similar. Sorry to bear the news lol

To note as a real-world example, in this case I looked at it because a friend was trying to work with the crate `bevy_hanabi`.

Not sure if the issue is reproducible with any stable version of that crate, but I did confirm it exists at this commit: https://github.com/djeedai/bevy_hanabi/tree/db782db43a03e8dbea41043f832fd5d5e468406f

In there, rust-analyzer produces this same false positive error in the file `examples/expr.rs` under basically the same circumstance.

**Comment by z-jxy** [comment]
Also having this issue https://github.com/rust-lang/rust-analyzer/issues/21668

**Comment by wellscrosby** [comment]
> Ooh, yeah, that does look very similar. Sorry to bear the news lol
> 
> To note as a real-world example, in this case I looked at it because a friend was trying to work with the crate `bevy_hanabi`.
> 
> Not sure if the issue is reproducible with any stable version of that crate, but I did confirm it exists at this commit: https://github.com/djeedai/bevy_hanabi/tree/db782db43a03e8dbea41043f832fd5d5e468406f
> 
> In there, rust-analyzer produces this same false positive error in the file `examples/expr.rs` under basically the same circumstance.

I'm having this exact issue with bevy-hanabi due to it having a pub struct Gradient and bevy_ui having a pub struct Gradient. Cargo check doesn't have the error, only rust-analyzer.
